src/vcpkg/base/diagnostics.cpp

```diff
@@ -0,0 +1,197 @@
+            }
+            break;
+            case DiagKind::Warning:
+            {
+                sink.print(Color::none, buf);
+                sink.print(Color::warning, "warning");
+                buf.assign(ColonSpace.data(), ColonSpace.size());
+            }
+            break;
+            case DiagKind::Note: buf.append(NotePrefix.data(), NotePrefix.size()); break;
+            default: Checks::unreachable(VCPKG_LINE_INFO);
+        }
+
+        buf.append(m_message.data());
+        buf.push_back('\n');
+        sink.print(Color::none, buf);
+    }
+    std::string DiagnosticLine::to_string() const
+    {
+        std::string result;
+        this->to_string(result);
+        return result;
+    }
+    void DiagnosticLine::to_string(std::string& target) const
+    {
+        append_file_prefix(target, m_origin, m_position);
+        append_kind_prefix(target, m_kind);
+        target.append(m_message.data());
+    }
+
+    LocalizedString DiagnosticLine::to_json_reader_string(const std::string& path, const LocalizedString& type) const
+    {
+        std::string result;
+        append_file_prefix(result, m_origin, m_position);
+        append_kind_prefix(result, m_kind);
+        result.append(path);
+        result.append(" (");
+        result.append(type.data());
+        result.append(") ");
+        result.append(m_message.data());
+        return LocalizedString::from_raw(result);
+    }
+
+    void BufferedDiagnosticContext::report(const DiagnosticLine& line)
+    {
+        std::lock_guard lck{m_mtx};
+        lines.push_back(line);
+    }
+
+    void BufferedDiagnosticContext::report(DiagnosticLine&& line)
+    {
+        std::lock_guard lck{m_mtx};
+        lines.push_back(std::move(line));
+    }
+    void BufferedDiagnosticContext::print(MessageSink& sink) const
+    {
+        for (auto&& line : lines)
+        {
+            line.print(sink);
+        }
+    }
+    // Converts this message into a string
+    // Prefer print() if possible because it applies color
+    // Not safe to use in the face of concurrent calls to report()
+    std::string BufferedDiagnosticContext::to_string() const
+    {
+        std::string result;
+        this->to_string(result);
+        return result;
+    }
+    void BufferedDiagnosticContext::to_string(std::string& target) const
+    {
+        auto first = lines.begin();
+        const auto last = lines.end();
+        if (first == last)
+        {
+            return;
+        }
+
+        for (;;)
+        {
+            first->to_string(target);
+            if (++first == last)
+            {
+                return;
+            }
+
+            target.push_back('\n');
+        }
+    }
+} // namespace vcpkg
+
+namespace
+{
+    struct ConsoleDiagnosticContext : DiagnosticContext
+    {
+        std::mutex mtx;
+        virtual void report(const DiagnosticLine& line) override
+        {
+            std::lock_guard<std::mutex> lck(mtx);
```

Comment:
>I think that if we are going to do synchronization here, this should be a batch interface.

There is no reason a batch interface could not be added. I did not add one because this PR was already big enough, and the addition of such a member doesn't change the structure of that already here. None of the APIs here desired a batch interface, so I didn't add one.

>This synchronization implies that it's usually correct to interleave lines coming from multiple threads. I suspect that's almost never correct

I believe it is almost always correct. (In fact, this belief is half the reason to open this can of worms in the first place, and that nothing here needed one added argues for this)

It was correct for all users of ParserBase herein.

>I'd honestly be more in favor of some system which crashes the process in debug mode if you try to call this function concurrently.

Debug mode functionally never happens for us, so that would be insufficient. I'm also unaware of practical ways to write an assert like that that would be better than an uncontended lock acquisition. You'd need something like a CAS to say "this thread is touching the thing", and a store afterwards to say that it is no longer touching the thing, which is exactly what an uncontended mutex lock/unlock is.